tests/components/apache_kafka/test_apache_kafka.py

```diff
@@ -0,0 +1,181 @@
+"""The tests for the Apache Kafka component."""
+from collections import namedtuple
+
+import pytest
+
+import homeassistant.components.apache_kafka as apache_kafka
+from homeassistant.const import STATE_ON
+from homeassistant.setup import async_setup_component
+
+from tests.async_mock import patch
+
+APACHE_KAFKA_PATH = "homeassistant.components.apache_kafka"
+PRODUCER_PATH = f"{APACHE_KAFKA_PATH}.AIOKafkaProducer"
+MIN_CONFIG = {
+    "ip_address": "localhost",
+    "port": 8080,
+    "topic": "topic",
+}
+FilterTest = namedtuple("FilterTest", "id should_pass")
+MockKafkaClient = namedtuple("MockKafkaClient", "init start send_and_wait")
+
+
+@pytest.fixture(name="mock_client")
+def mock_client_fixture():
+    """Mock the apache kafka client."""
```

Comment:
Untested, but something like this might work to avoid patching `__init__`

```
        mock_kafka = Mock(PRODUCER_PATH)
        send_and_wait = AsyncMock()
        type(mock_kafka).send_batch = send_and_wait

       ...
```

Reply:
So I tried this but it didn't seem to work, still getting errors. I also just tried a whole bunch of other options to just patch the non-init functions I need and nothing else is really seeming to work here. Either I get errors about trying to await on `MagicMock` objects or I get errors from within the implementation object because one or more of its methods are actually being called instead of being mocked.

The only one that seems to come close to being successful is if I do this
```python
with patch(PRODUCER_PATH, autospec=True) as client:
    yield client
```
This still doesn't work but the errors I get are assertion errors rather then runtime errors:
```
FAILED test_apache_kafka.py::test_minimal_config[pyloop] - AttributeError: Mock object has no attribute 'called_once'
FAILED test_apache_kafka.py::test_full_config[pyloop] - AttributeError: Mock object has no attribute 'called_once'
FAILED test_apache_kafka.py::test_allowlist[pyloop] - AssertionError: Expected 'send_and_wait' to have been called once. Called 0 times.
FAILED test_apache_kafka.py::test_denylist[pyloop] - AssertionError: Expected 'send_and_wait' to have been called once. Called 0 times.
FAILED test_apache_kafka.py::test_filtered_allowlist[pyloop] - AssertionError: Expected 'send_and_wait' to have been called once. Called 0 times.
FAILED test_apache_kafka.py::test_filtered_denylist[pyloop] - AssertionError: Expected 'send_and_wait' to have been called once. Called 0 times.
```
So it suggests to me that this object is being mocked (since I don't get internal apache errors) and the async status of methods is being retained (since I don't get can't await MagicMock errors) but that I'm not properly properly spec'cing these methods. At least I think so, I'm not very familiar with speccing a mock from a class so I'm not totally sure what is happening.

Reply:
Let's leave it for now.  If someone suggests a better way, we can adjust it later.